client/blockchain-service/src/handler.rs

```diff
@@ -902,4 +934,45 @@ impl BlockchainService {
         };
         current_tick == &last_tick_provided.saturating_add(provider_challenge_period)
     }
+
+    fn check_pending_forest_root_writes(&mut self) {
+        if let Some(mut rx) = self.forest_root_write_lock.take() {
+            // TODO: consider adding a timeout here to prevent a task from blocking the runtime.
+            match rx.try_recv() {
+                // If the channel is empty, means we still need to wait for the current task to finish.
+                Err(TryRecvError::Empty) => self.forest_root_write_lock = Some(rx),
+                _ => {}
+            }
+        }
+
+        // If we have a task writing to the runtime, we don't want to start another one.
+        if self.forest_root_write_lock.is_some() {
+            return;
+        }
+
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        self.forest_root_write_lock = Some(rx);
+
+        let forest_root_write_tx = Arc::new(Mutex::new(Some(tx)));
```

Comment:
*(note from author)*

This is an `Arc<Mutex<Option<T>>>` (in this case `T=oneshot::Sender<()>`) instead of just `T` so that we can keep using the current actors event bus (emit) which requires `Clone` on the event. Clone is required because there is no constraint on the number of listeners that can subscribe to the event (and each is guaranteed to receive all emitted events).

Reply:
That's probably worth commenting in the code.